certbot-postfix/certbot_postfix/util.py

```diff
@@ -0,0 +1,254 @@
+    """Ensures an executable with the given name is available.
+
+    If an executable isn't found for the given path or name, extra
+    directories are added to the user's PATH to help find system
+    utilities that may not be available in the default cron PATH.
+
+    :param str exe: executable path or name
+    :param str message: Error message to print.
+
+    :raises .NoInstallationError: when the executable isn't found
+
+    """
+    if message is None:
+        message = "Cannot find executable '{0}'.".format(exe)
+    if not (certbot_util.exe_exists(exe) or plugins_util.path_surgery(exe)):
+        raise errors.NoInstallationError(message)
+
+def report_master_overrides(name, overrides, acceptable_overrides=None):
+    """If the value for a parameter |name| is overridden by other services,
+    report a warning to notify the user.
+
+    :param str name: The name of the parameter that is being overridden.
+    :param list overrides: The values that other services are setting for |name|.
+        Each override is a tuple: (service name, value)
+    :param list acceptable_overrides: Override values that are acceptable. For instance, if
+        another service is overriding our parameter with a more secure option, we don't have
+        to warn. If this is set to None, warnings are reported for *all* overrides!
+    """
+    error_string = ""
+    for override in overrides:
+        service, value = override
+        # If this override is acceptable:
+        if acceptable_overrides is not None and \
+            is_acceptable_value(name, value, acceptable_overrides):
+            continue
+        error_string += "  {0}: {1}\n".format(service, value)
+    if len(error_string) > 0:
+        raise errors.PluginError("{0} is overridden with less secure options by the "
+             "following services in master.cf:\n".format(name) + error_string)
+
+def is_acceptable_value(parameter, value, acceptable):
+    """ Returns whether the `value` for this `parameter` is acceptable,
+    given a string or tuple `acceptable`
+    """
+    # If it's a tuple, there's multiple acceptable options.
+    # Only set a param if it's not acceptable.
+    if isinstance(acceptable, tuple):
+        return value in acceptable
+    # Check if param value is a comma-separated list of protocols.
+    elif 'protocols' in parameter:
+        return _has_acceptable_tls_versions(value)
+    # Otherwise, just check whether the value is equal to acceptable.
+    return value == acceptable
+
+def _has_acceptable_tls_versions(parameter_string):
+    """
+    Checks to see if the comma-separated list of TLS protocols to exclude is acceptable.
+    Sample string: "!SSLv2, !SSLv3"
+    """
+    bad_versions = list(constants.TLS_VERSIONS)
+    for version in constants.ACCEPTABLE_TLS_VERSIONS:
+        del bad_versions[bad_versions.index(version)]
+    for bad_version in bad_versions:
+        if "!" + bad_version not in parameter_string:
```

Comment:
What if `parameter_string` lists protocols to include rather than those to exclude? This seems allowed according to docs like http://www.postfix.org/postconf.5.html#smtpd_tls_protocols.